src/board/v3/nap/track_channel.c

```diff
@@ -28,74 +29,105 @@
+  (NAP_TRACK_CODE_PHASE_UNITS_PER_CHIP / (double)TRACK_SAMPLE_FREQ)
 
-#define SPACING_HALF_CHIP         ((u16)(TRACK_SAMPLE_FREQ / GPS_CA_CHIPPING_RATE) / 2)
+#define NAP_TRACK_CODE_PHASE_UNITS_PER_CHIP       \
+  ((u64)1 << NAP_TRACK_CODE_PHASE_FRACTIONAL_WIDTH)
 
-BSEMAPHORE_DECL(timing_strobe_sem, TRUE);
+#define SPACING_HALF_CHIP ((u16)(TRACK_SAMPLE_FREQ / GPS_CA_CHIPPING_RATE) / 2)
 
-static struct {
-  u32 code_phase;
-  u32 len;
+static struct nap_ch_state {
+  u32 code_phase;   /**< Fractional part of code phase. */
 } nap_ch_state[NAP_MAX_N_TRACK_CHANNELS];
 
-static u32 calc_length_samples(u8 codes, s32 cp_start, u32 cp_rate)
+static u32 calc_length_samples(u8 codes, u32 cp_start_frac_units,
+                               u32 cp_rate_units)
 {
-  u16 chips = codes * 1023;
-  u64 cp_units = (1ULL << NAP_TRACK_CODE_PHASE_FRACTIONAL_WIDTH) * chips
-                 - cp_start;
-  u32 samples = cp_units / cp_rate;
+  u64 cp_end_units = codes * 1023 * NAP_TRACK_CODE_PHASE_UNITS_PER_CHIP;
+  /* cp_start_frac_units is reinterpreted as a signed value. This works
+   * because NAP_TRACK_CODE_PHASE_FRACTIONAL_WIDTH is equal to 32 */
+  u64 cp_units = cp_end_units - (s32)cp_start_frac_units;
+  u32 samples = cp_units / cp_rate_units;
   return samples;
 }
 
 void nap_track_init(u8 channel, gnss_signal_t sid, u32 ref_timing_count,
-                   float carrier_freq, float code_phase)
+                    float carrier_freq, float code_phase)
 {
-  u32 now = NAP->TIMING_COUNT;
-  u32 track_count = now + 200000;
-  double cp = propagate_code_phase(code_phase, carrier_freq,
-                                  track_count - ref_timing_count);
-
-  /* Contrive for the timing strobe to occur at or close to a
-   * PRN edge (code phase = 0) */
-  track_count += (SAMPLE_FREQ/GPS_CA_CHIPPING_RATE) * (1023.0-cp) *
-                 (1.0 + carrier_freq / GPS_L1_HZ);
-
-  u8 prn = sid.sat - 1;
-  NAP->TRK_CH[channel].CONTROL = prn << NAP_TRK_CONTROL_SAT_Pos;
+  assert(sid.code == CODE_GPS_L1CA);
+
+  nap_trk_regs_t *t = &NAP->TRK_CH[channel];
+  struct nap_ch_state *s = &nap_ch_state[channel];
+  memset(s, 0, sizeof(*s));
+
+  u8 prn = sid.sat - GPS_FIRST_PRN;
+  t->CONTROL = prn << NAP_TRK_CONTROL_SAT_Pos;
+
   /* We always start at zero code phase */
-  NAP->TRK_CH[channel].CODE_INIT_INT = 0;
-  NAP->TRK_CH[channel].CODE_INIT_FRAC = 0;
-  NAP->TRK_CH[channel].CODE_INIT_G1 = 0x3ff;
-  NAP->TRK_CH[channel].CODE_INIT_G2 = 0x3ff;
+  t->CODE_INIT_INT = 0;
+  t->CODE_INIT_FRAC = 0;
+  t->CODE_INIT_G1 = 0x3ff;
+  t->CODE_INIT_G2 = 0x3ff;
+
+  t->SPACING = (SPACING_HALF_CHIP << NAP_TRK_SPACING_OUTER_Pos) |
+               (SPACING_HALF_CHIP << NAP_TRK_SPACING_INNER_Pos);
+
+  double cp_rate = (1.0 + carrier_freq / GPS_L1_HZ) * GPS_CA_CHIPPING_RATE;
+  nap_track_update(channel, carrier_freq, cp_rate, 0, 0);
 
-  NAP->TRK_CH[channel].SPACING = (SPACING_HALF_CHIP << 16) | SPACING_HALF_CHIP;
+  /* First integration is one sample short */
```

Comment:
Yes, but I don't think this is very important now that the phases are read back from the NAP.